nautobot/ui/src/components/Table.js

```diff
@@ -1,48 +1,99 @@
+import { useEffect, useState, useMemo, useCallback } from "react";
 import {
-    Checkbox,
-    Table,
-    Thead,
-    Tr,
-    Th,
-    Tbody,
-    Td,
+    TableRenderer,
+    useTableRenderer,
+    createColumnHelper,
+    Button,
+    EditIcon,
 } from "@nautobot/nautobot-ui";
 
 import TableItem from "@components/TableItem";
 
+function getTableItemLink(idx, obj) {
+    if (idx === 0) {
+        // TODO: ui-schema should be providing the name of the field used to describe the object (name)
+        return window.location.pathname + obj.id + "/";
+    }
+    if (typeof obj !== "object" || !obj || !obj.url) {
+        return null;
+    }
+    // Remove domain + /api prefix
+    return obj.url.replace(window.location.origin + "/api", "");
+}
+
+function createTableColumn(tableHeaders, columnHelper) {
+    return tableHeaders.map(({ key, title }, idx) =>
+        columnHelper.accessor(key, {
+            cell: (props) => {
+                // Get the column data from the object
+                // e.g from {"status": {"display": "Active"}, "id": ....} get => {"display": "Active"}
+                const column_data =
+                    idx === 0
+                        ? props.row.original
+                        : props.row.original[props.column.id];
+                return (
+                    <TableItem
+                        name={key}
+                        obj={props.getValue()}
+                        url={getTableItemLink(idx, column_data)}
+                    />
+                );
+            },
+            header: title,
+        })
+    );
+}
+
+function renderActionButton({ cellContext }) {
+    return (
+        <Button
+            leftIcon={<EditIcon size="sm" />}
+            size="xs"
+            variant="table"
+            onClick={() => alert(`Clicked ${cellContext.row.original.name}!`)}
+        />
+    );
+}
+
 // A standard Nautobot table. This _may_ be beneficial to move into nautobot-ui
-export default function NautobotTable({ data, headers }) {
+export default function NautobotTable({
```

Comment:
Sorry for bothering you with this but can I get some comments on what some of the react hooks are doing?

Reply:
Im attaching link to the documentation for each hook:

1. [useMemo](https://react.dev/reference/react/useMemo)
2. [useCallback](https://react.dev/reference/react/useCallback)

Its basically a copy paste code from what was in ObjectRetrieve.js initially